cli/genpolicy/genpolicy.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2024 Edgeless Systems GmbH
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package genpolicy
+
+import (
+	"context"
+	_ "embed"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+
+	"github.com/edgelesssys/contrast/internal/embedbin"
+	"github.com/edgelesssys/contrast/node-installer/platforms"
+)
+
+var (
+	//go:embed assets/genpolicy
+	genpolicyBin []byte
+	//go:embed assets/genpolicy-settings.json
+	defaultGenpolicySettings []byte
+	//go:embed assets/genpolicy-rules.rego
+	aksCloudHypervisorSNPRules []byte
+	//go:embed assets/allow-all.rego
+	permissiveRules []byte
+)
+
+// Runner is a wrapper around the genpolicy tool.
+//
+// Create an instance with New(), optionally configure it using the exported
+// fields, call Run() to execute the tool, and call Teardown() afterwards to
+// clean up temporary files.
+type Runner struct {
+	Rules     []byte
+	Settings  []byte
+	CacheFile string
+
+	once      sync.Once
+	genpolicy embedbin.Installed
+	settings  embedbin.Installed
+	rules     embedbin.Installed
+}
+
+// New creates a new Runner with the default configuration for the requested platform.
+func New(platform platforms.Platform) (*Runner, error) {
+	cacheDir, err := os.UserCacheDir()
+	if err != nil {
+		return nil, fmt.Errorf("getting cache dir: %w", err)
+	}
+	runner := &Runner{
+		Settings:  defaultGenpolicySettings,
+		CacheFile: filepath.Join(cacheDir, "contrast", "layer-cache.json"),
+	}
+	switch platform {
+	case platforms.AKSCloudHypervisorSNP:
+		runner.Rules = aksCloudHypervisorSNPRules
+	default:
+		// TODO(burgerdev): use real rules for supported platforms.
+		runner.Rules = permissiveRules
+	}
+
+	return runner, nil
+}
+
+// Run runs the tool on the given yaml.
+//
+// Run can be called more than once, but if the first run fails during
+// initialization subsequent runs won't succeed.
+func (r *Runner) Run(ctx context.Context, yamlPath string, logger *slog.Logger) error {
+	var outerErr error
+	r.once.Do(func() {
+		var err error
+		e := embedbin.New()
+		r.genpolicy, err = e.Install("", genpolicyBin)
+		outerErr = errors.Join(outerErr, err)
+		r.settings, err = e.Install("", r.Settings)
+		outerErr = errors.Join(outerErr, err)
+		r.rules, err = e.Install("", r.Rules)
+		outerErr = errors.Join(outerErr, err)
+	})
+	if outerErr != nil {
+		return fmt.Errorf("creating temp files: %w", outerErr)
+	}
```

Comment:
AFAICT this error will only be caught on the first `Run` call, on subsequent calls no error is generated and so the code down below will fail with a null-pointer exception. 

As of right now, AFAICT there's no scenario where this could happen, but still, it'd be nice to return a proper error message instead of the null-pointer one.

Reply:
I rewrote this part, don't need the `sync.Once` anymore.